containerize/process_log.go

```diff
@@ -0,0 +1,100 @@
+package containerize
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"sync"
+	"teleport-exec/filestream"
+)
+
+//-------------------------------------------------------------------------------------------------
+type ProcessLog struct {
+	command_id  string
+	fd          *os.File
+	readers     map[*filestream.FileStream]bool
+	readersLock sync.Mutex
+}
+
+func NewProcessLog(command_id string) (*ProcessLog, error) {
+	file_name := "/tmp/command-" + command_id + ".out"
+
+	fd, err := os.Create(file_name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create a log file for command %s (%s): %w", command_id, file_name, err)
+	}
+
+	l := &ProcessLog{
+		command_id: command_id,
+		readers:    make(map[*filestream.FileStream]bool),
+		fd:         fd,
+	}
+
+	return l, nil
+}
+
+//-------------------------------------------------------------------------------------------------
+func (l *ProcessLog) FileName() string {
+	return l.fd.Name()
+}
+
+// Closes the log and deletes the log file
+func (l *ProcessLog) Close() (err error) {
+	l.readersLock.Lock()
+	defer l.readersLock.Unlock()
+
+	// Close the log and delete the log file
+	err = l.fd.Close()
+	if err == nil {
+		err = os.Remove(l.FileName())
+	}
+
+	// Close all readers
+	for reader := range l.readers {
+		err = reader.Close()
+		if err != nil {
+			return err
+		}
+	}
+
+	return err
+}
+
+//-------------------------------------------------------------------------------------------------
+// Tells all readers to stop waiting for more content since the log is complete (command is done)
+func (l *ProcessLog) LogComplete() {
+	l.readersLock.Lock()
+	defer l.readersLock.Unlock()
+
+	for reader := range l.readers {
+		reader.DisableTail()
+	}
+}
+
+//-------------------------------------------------------------------------------------------------
+// Returns a new file reader for the log.
+func (l *ProcessLog) NewLogStream(ctx context.Context, tail bool) (stream *filestream.FileStream, err error) {
```

Comment:
Requiring a flag to enable tail mode seems fragile. Any way we can do without it?

Reply:
Another option I see is having an alternative implementation of the stream, which simply does not support tailing (a very thin wrapper around `io.File` or something like that) and picking one implementation over the other depending on the state of the command we're trying to observe. But the filestream implementation still needs the tail flag support or some other mechanism for notifying blocked streams that the log is complete.